astropy/cosmology/connect.py

```diff
@@ -178,10 +194,74 @@ class CosmologyFromFormat(io_registry.UnifiedReadWrite):
         `~astropy.cosmology.Cosmology` corresponding to ``obj`` contents.
     """
 
-    def __init__(self, instance, cosmo_cls):
+    def __init__(self, instance: Cosmology, cosmo_cls: type[Cosmology]) -> None:
         super().__init__(instance, cosmo_cls, "read", registry=convert_registry)
 
-    def __call__(self, obj, *args, format=None, **kwargs):
+    # ===============================================================
+    # __call__ overloads
+    # note: format: ... | None means the format can be auto-detected from the input.
+
+    @overload
+    def __call__(
+        self,
+        obj: _CosmoT,
+        *args: Any,
+        format: Literal["astropy.cosmology"] | None,
+        **kwargs: Any,
+    ) -> _CosmoT:
+        ...
+
+    @overload
+    def __call__(
+        self,
+        obj: _CosmologyModel,
+        *args: Any,
+        format: Literal["astropy.model"] | None,
+        **kwargs: Any,
+    ) -> Cosmology:
+        ...
+
+    @overload
+    def __call__(
+        self, obj: Row, *args: Any, format: Literal["astropy.row"] | None, **kwargs: Any
+    ) -> Cosmology:
+        ...
+
+    @overload
+    def __call__(
+        self,
+        obj: Table,
+        *args: Any,
+        format: Literal["astropy.table"] | None,
+        **kwargs: Any,
+    ) -> Cosmology:
+        ...
+
+    @overload
+    def __call__(
+        self,
+        obj: Mapping[str, Any],
+        *args: Any,
+        format: Literal["mapping"] | None,
+        **kwargs: Any,
+    ) -> Cosmology:
+        ...
+
+    @overload
+    def __call__(
+        self, obj: str, *args: Any, format: Literal["yaml"], **kwargs: Any
+    ) -> Cosmology:
+        ...
+
+    @overload
+    def __call__(
+        self, obj: Any, *args: Any, format: str | None = None, **kwargs: Any
+    ) -> Cosmology:
+        ...
```

Comment:
Do we need this many overloads for `CosmologyFromFormat`? There seem to be only two cases: `obj: _CosmoT` and `obj: Any`. It looks to me like the base case with `obj: Any` and `format: str | None` means that type checkers will not warn us even if we would use e.g. `obj: Mapping` together with `format="yaml"`.

Reply:
Mypy didn't complain about any overlapping overloads.
And my IDE gives a very helpful full rundown of the overloads, which is a very explicit list of all the builtin formats + the more generic user-registration option. With the IDE you can see on a hover-over that for a string input we can't auto-identify that it's YAML.

![CleanShot 2024-01-16 at 17 07 24](https://github.com/astropy/astropy/assets/8949649/dc3b002e-6695-4154-8ed6-f0d9ce4239d9)

Reply:
If you try to call this function with `obj: Row` and `format="astropy.table"` then does your IDE highlight that as a mistake?

Reply:
Unfortunately it doesn't. And the most obvious solution of a type negation, e.g. `str | None | Not[Literal[....]]` isn't coming any time soon (https://github.com/python/typing/issues/801)

Reply:
The overloads can be helpful at the IDE level, but not for catching bugs via static analysis.